File: integration_test/src/java/org/apache/heron/integration_test/core/TestTopologyBuilder.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.heron.integration_test.core;

import java.io.Serializable;
import java.lang.reflect.InvocationTargetException;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;

import org.apache.heron.api.Config;
import org.apache.heron.api.HeronTopology;
import org.apache.heron.api.bolt.IRichBolt;
import org.apache.heron.api.bolt.IStatefulWindowedBolt;
import org.apache.heron.api.bolt.IWindowedBolt;
import org.apache.heron.api.bolt.StatefulWindowedBoltExecutor;
import org.apache.heron.api.bolt.WindowedBoltExecutor;
import org.apache.heron.api.generated.TopologyAPI;
import org.apache.heron.api.spout.IRichSpout;
import org.apache.heron.api.topology.BoltDeclarer;
import org.apache.heron.api.topology.IStatefulComponent;
import org.apache.heron.api.topology.SpoutDeclarer;
import org.apache.heron.api.topology.TopologyBuilder;

public class TestTopologyBuilder extends TopologyBuilder {
  private static final int DEFAULT_EXECUTION_COUNT = 10;

  // This variable will be used as input variable for constructor of our aggregator bolt
  // This will determine the location of where our output is directed
  // Could be URL, file location, etc.
  private final String outputLocation;
  private final String stateLocation;
  private final String stateUpdateToken;
  private final SpoutWrapperType spoutWrapperType;

  // The structure of the topologyBlr - a graph directed from children to parents
  private final Map<String, TopologyAPI.Bolt.Builder> bolts = new HashMap<>();
  private final Map<String, TopologyAPI.Spout.Builder> spouts = new HashMap<>();
  private final Map<String, HashSet<String>> prev = new HashMap<>();
  // By default, terminalBoltClass will be AggregatorBolt, which writes to specified HTTP server
  private String terminalBoltClass = "org.apache.heron.integration_test.core.AggregatorBolt";

  public enum SpoutWrapperType {
    DEFAULT,
    TWO_PHASE,
    EMIT_UNTIL
  }

  public TestTopologyBuilder(String outputLocation) {
    this(outputLocation, null, null, SpoutWrapperType.DEFAULT);
  }

  public TestTopologyBuilder(String outputLocation,
                             String stateLocation,
                             String stateUpdateToken,
                             SpoutWrapperType spoutWrapperType) {
    this.outputLocation = outputLocation;
    this.stateLocation = stateLocation;
    this.stateUpdateToken = stateUpdateToken;
    this.spoutWrapperType = spoutWrapperType;
  }

  @Override
  public BoltDeclarer setBolt(String id, IRichBolt bolt, Number parallelismHint) {
    return setBolt(id, bolt, parallelismHint, true);
  }

  public BoltDeclarer setBolt(String id, IRichBolt bolt, Number parallelismHint, boolean ackAuto) {
    if (bolt instanceof IStatefulComponent) {
      return super.setBolt(id, new StatefulIntegrationTestBolt<>(bolt, ackAuto), parallelismHint);
    } else {
      return super.setBolt(id, new IntegrationTestBolt(bolt, ackAuto), parallelismHint);
    }
  }

  public BoltDeclarer setBolt(String id, IWindowedBolt bolt,
                              Number parallelismHint, boolean ackAuto) throws
      IllegalArgumentException {
    return setBolt(id, new WindowedBoltExecutor(bolt), parallelismHint, ackAuto);
  }

  @SuppressWarnings("rawtypes")
  public <K extends Serializable, V extends Serializable> BoltDeclarer setBolt(
      String id, IStatefulWindowedBolt<K, V> bolt, Number parallelismHint, boolean ackAuto) throws
      IllegalArgumentException {
    return setBolt(id, new StatefulWindowedBoltExecutor<K, V>(bolt), parallelismHint, ackAuto);
  }

  @Override
  public SpoutDeclarer setSpout(String id, IRichSpout spout, Number parallelismHint) {
    return setSpout(id, spout, parallelismHint, DEFAULT_EXECUTION_COUNT);
  }

  // A method allows user to define the maxExecutionCount of the spout
  // To be compatible with earlier Integration Test Framework
  public SpoutDeclarer setSpout(String id, IRichSpout spout,
                                Number parallelismHint, int maxExecutionCount) {
    String topologyStartedUrl = null;
    String tuplesEmittedUrl = null;
    String topologyUpdateUrl = null;
    if (stateLocation != null) {
      topologyStartedUrl = stateLocation + "_topology_started";
      tuplesEmittedUrl = stateLocation + "_tuples_emitted";
      topologyUpdateUrl = stateLocation + "_" + stateUpdateToken;
    }
    IntegrationTestSpout wrappedSpout;
    switch (spoutWrapperType) {
      case TWO_PHASE:
        wrappedSpout = new MultiPhaseTestSpout(spout, maxExecutionCount, 2,
            new HttpGetCondition(topologyUpdateUrl), topologyStartedUrl);
        break;
      case EMIT_UNTIL:
        wrappedSpout = new EmitUntilConditionTestSpout(spout,
            new HttpGetCondition(topologyUpdateUrl), topologyStartedUrl, tuplesEmittedUrl);
        break;
      case DEFAULT:
      default:
        if (spout instanceof IStatefulComponent) {
          wrappedSpout
              = new StatefulIntegrationTestSpout<>(spout, maxExecutionCount, topologyStartedUrl);
        } else {
          wrappedSpout = new IntegrationTestSpout(spout, maxExecutionCount, topologyStartedUrl);
        }
    }

    return setSpout(id, wrappedSpout, parallelismHint);
  }

  private SpoutDeclarer setSpout(String id, IntegrationTestSpout itSpout, Number parallelismHint) {
    return super.setSpout(id, itSpout, parallelismHint);
  }

  public void setTerminalBoltClass(String terminalBoltClass) {
    this.terminalBoltClass = terminalBoltClass;
  }

  // By default, will use AggregatorBolt, which writes to HTTP Server and takes URL as input
  @Override
  public HeronTopology createTopology() {
    // We will add the aggregation_bolt to be serialized
    final String AGGREGATOR_BOLT = "__integration_test_aggregator_bolt";
    BaseBatchBolt aggregatorBolt;
    try {
      // Terminal Bolt will be initialized using reflection, based on the value of terminal bolt
      // class.
      // class should be built on top of BaseBatchBolt abstract class, and can be changed using
      // setTerminalBolt function
      aggregatorBolt =
          (BaseBatchBolt) Class.forName(terminalBoltClass).getConstructor(String.class)
              .newInstance(this.outputLocation);
    } catch (NoSuchMethodException e) {
      throw new RuntimeException(e + " Terminal Bolt class must have a single String constructor.");
    } catch (InstantiationException e) {
      throw new RuntimeException(e + " Terminal bolt class could not be instantiated.");
    } catch (IllegalAccessException e) {
      throw new RuntimeException(e + " Terminal Bolt class constructor is not accessible.");
    } catch (InvocationTargetException e) {
      throw new RuntimeException(e + " Terminal Bolt class constructor could not be invoked.");
    } catch (ClassNotFoundException e) {
      throw new RuntimeException(e + " Terminal Bolt class must be a class path.");
    }
    setBolt(AGGREGATOR_BOLT, aggregatorBolt, 1);

    // We get the user-defined TopologyAPI.Topology.Builder
    TopologyAPI.Topology.Builder topologyBlr =
        super.createTopology().
            setConfig(new Config()).
            setName("").
            setState(TopologyAPI.TopologyState.RUNNING).
            getTopology().toBuilder();

    // Clear unnecessary fields to make the state of TopologyAPI.Topology.Builder clean
    topologyBlr.clearTopologyConfig().clearName().clearState();

    for (TopologyAPI.Spout.Builder spout : topologyBlr.getSpoutsBuilderList()) {
      String name = spout.getComp().getName();
      spouts.put(name, spout);
    }

    // We will build the structure of the topologyBlr - a graph directed from children to parents,
    // by looking only on bolts, since spout will not have parents
    for (TopologyAPI.Bolt.Builder bolt : topologyBlr.getBoltsBuilderList()) {
      String name = bolt.getComp().getName();
      bolts.put(name, bolt);

      if (name.equals(AGGREGATOR_BOLT)) {
        // We should not consider aggregator bolt in the topology's graph or terminal components
        // since it is not user-defined
        continue;
      }

      // To get the parent's component to construct a graph of topology structure
      for (TopologyAPI.InputStream inputStream : bolt.getInputsList()) {
        String parent = inputStream.getStream().getComponentName();
        if (prev.containsKey(name)) {
          prev.get(name).add(parent);
        } else {
          HashSet<String> parents = new HashSet<String>();
          parents.add(parent);
          prev.put(name, parents);
        }
      }
    }

    // To find the terminal bolts defined by users and link them with "AggregatorBolt"
    // First, "it" of course needs upstream component, we don't want the isolated bolt
    HashSet<String> terminals = new HashSet<>();
    // Second, "it" should not exists in the prev.valueSet, which means, it has no downstream
    HashSet<String> nonTerminals = new HashSet<>();
    for (HashSet<String> set : prev.values()) {
      nonTerminals.addAll(set);
    }
    // Here we iterate bolt in prev.keySet() rather than bolts.keySet() due to we don't want
    // a isolated bolt, including AggregatorBolt
    for (String bolt : prev.keySet()) {
      if (!nonTerminals.contains(bolt)) {
        terminals.add(bolt);
      }
    }
    // We will also consider the cases with spouts without children
    for (String spout : spouts.keySet()) {
      if (!nonTerminals.contains(spout)) {
        terminals.add(spout);
      }
    }

    // Now first, we will add all grouping to components
    for (String child : prev.keySet()) {
      for (String parent : prev.get(child)) {
        addAllGrouping(child, parent, Constants.INTEGRATION_TEST_CONTROL_STREAM_ID);
      }
    }

    // Then we need to connect aggregator bolt with user's terminal components
    // We could use any grouping but for convenience we would use allGrouping here
    for (String t : terminals) {
      List<TopologyAPI.OutputStream> osList;
      if (bolts.get(t) != null) {
        osList = bolts.get(t).getOutputsList();
      } else {
        osList = spouts.get(t).getOutputsList();
      }
      for (TopologyAPI.OutputStream os : osList) {
        addAllGrouping(AGGREGATOR_BOLT, t, os.getStream().getId());
      }
    }

    // We wrap it to the new topologyBuilder
    return new HeronTopology(topologyBlr);
  }

  /**
   * Given a child component and its upstream component, add the allGrouping between them with
   * given streamId
   *
   * @param child the child component id
   * @param parent the upstream component id
   * @param streamId the stream id
   */
  protected void addAllGrouping(String child, String parent, String streamId) {
    TopologyAPI.InputStream.Builder builder = TopologyAPI.InputStream.newBuilder();
    builder.setStream(TopologyAPI.StreamId.newBuilder().setId(streamId).setComponentName(parent));
    builder.setGtype(TopologyAPI.Grouping.ALL);

    bolts.get(child).addInputs(builder);
  }
}

File: integration_test/src/python/integration_test/core/test_topology_builder.py
#!/usr/bin/env python
# -*- encoding: utf-8 -*-

#  Licensed to the Apache Software Foundation (ASF) under one
#  or more contributor license agreements.  See the NOTICE file
#  distributed with this work for additional information
#  regarding copyright ownership.  The ASF licenses this file
#  to you under the Apache License, Version 2.0 (the
#  "License"); you may not use this file except in compliance
#  with the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
#  Unless required by applicable law or agreed to in writing,
#  software distributed under the License is distributed on an
#  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
#  KIND, either express or implied.  See the License for the
#  specific language governing permissions and limitations
#  under the License.

'''integration test topology builder'''
import copy
import heronpy.api.api_constants as api_constants
from heronpy.api.stream import Stream, Grouping
from heronpy.api.topology import TopologyBuilder, Topology, TopologyType
from ..core import constants as integ_const
from .aggregator_bolt import AggregatorBolt
from .integration_test_spout import IntegrationTestSpout
from .integration_test_bolt import IntegrationTestBolt

class TestTopologyBuilder(TopologyBuilder):
  """Topology Builder for integration tests

  Given spouts and bolts will be delegated by IntegrationTestSpout and IntegrationTestBolt
  classes respectively.
  """
  TERMINAL_BOLT_NAME = '__integration_test_aggregator_bolt'
  TERMINAL_BOLT_CLASS = AggregatorBolt
  DEFAULT_CONFIG = {api_constants.TOPOLOGY_DEBUG: True,
                    api_constants.TOPOLOGY_RELIABILITY_MODE:
                    api_constants.TopologyReliabilityMode.ATLEAST_ONCE,
                    api_constants.TOPOLOGY_PROJECT_NAME: "heron-integration-test"}
  def __init__(self, name, http_server_url):
    super(TestTopologyBuilder, self).__init__(name)
    self.output_location = "%s/%s" % (http_server_url, self.topology_name)
    self.set_config(self.DEFAULT_CONFIG)

    # map <name -> spout's component spec>
    self.spouts = {}
    # map <name -> bolt's component spec>
    self.bolts = {}
    # map <name -> set of parents>
    self.prev = {}

  def add_spout(self, name, spout_cls, par, config=None,
                optional_outputs=None, max_executions=None):
    """Add an integration_test spout"""
    user_spec = spout_cls.spec(name)
    spout_classpath = user_spec.python_class_path

    if hasattr(spout_cls, 'outputs'):
      user_outputs = spout_cls.outputs
    else:
      user_outputs = []

    if optional_outputs is not None:
      user_outputs.extend(optional_outputs)

    if config is None:
      _config = {}
    else:
      _config = copy.copy(config)

    if max_executions is not None:
      _config[integ_const.USER_MAX_EXECUTIONS] = max_executions

    test_spec = IntegrationTestSpout.spec(name, par, _config,
                                          user_spout_classpath=spout_classpath,
                                          user_output_fields=user_outputs)
    self.add_spec(test_spec)
    self.spouts[name] = test_spec
    return test_spec

  def add_bolt(self, name, bolt_cls, par, inputs, config=None, optional_outputs=None):
    """Add an integration_test bolt

    Only dict based inputs is supported
    """
    assert isinstance(inputs, dict)
    user_spec = bolt_cls.spec(name)
    bolt_classpath = user_spec.python_class_path

    if hasattr(bolt_cls, 'outputs'):
      user_outputs = bolt_cls.outputs
    else:
      user_outputs = []

    if optional_outputs is not None:
      user_outputs.extend(optional_outputs)

    if config is None:
      _config = {}
    else:
      _config = config

    test_spec = IntegrationTestBolt.spec(name, par, inputs, _config,
                                         user_bolt_classpath=bolt_classpath,
                                         user_output_fields=user_outputs)
    self.add_spec(test_spec)
    self.bolts[name] = test_spec
    return test_spec

  # pylint: disable=too-many-branches
  def create_topology(self):
    """Creates an integration_test topology class"""

    # first add the aggregation_bolt
    # inputs will be updated later
    aggregator_config = {integ_const.HTTP_POST_URL_KEY: self.output_location}
    self.add_bolt(self.TERMINAL_BOLT_NAME, self.TERMINAL_BOLT_CLASS, 1,
                  inputs={}, config=aggregator_config)

    # building a graph directed from children to parents, by looking only on bolts
    # since spouts don't have parents
    for name, bolt_spec in self.bolts.items():
      if name == self.TERMINAL_BOLT_NAME:
        continue

      bolt_protobuf = bolt_spec.get_protobuf()
      for istream in bolt_protobuf.inputs:
        parent = istream.stream.component_name
        if name in self.prev:
          self.prev[name].add(parent)
        else:
          parents = set()
          parents.add(parent)
          self.prev[name] = parents

    # Find the terminal bolts defined by users and link them with "AggregatorBolt".

    # set of terminal component names
    terminals = set()
    # set of non-terminal component names
    non_terminals = set()
    # 1. terminal bolts need upstream components, because we don't want isolated bolts
    # 2. terminal bolts should not exist in the prev.values(), meaning that no downstream
    for parent_set in list(self.prev.values()):
      non_terminals.update(parent_set)

    for bolt_name in list(self.prev.keys()):
      if bolt_name not in non_terminals:
        terminals.add(bolt_name)

    # will also consider the cases with spouts without children
    for spout_name in list(self.spouts.keys()):
      if spout_name not in non_terminals:
        terminals.add(spout_name)

    # add all grouping to components
    for child in list(self.prev.keys()):
      for parent in self.prev[child]:
        self._add_all_grouping(child, parent, integ_const.INTEGRATION_TEST_CONTROL_STREAM_ID)

    # then connect aggregator bolt with user's terminal components
    # terminal_outputs are output fields for terminals, list of either str or Stream
    for terminal in terminals:
      if terminal in self.bolts:
        terminal_outputs = self.bolts[terminal].outputs
      else:
        terminal_outputs = self.spouts[terminal].outputs

      # now get a set of stream ids
      stream_ids = ["default" if isinstance(out, str) else out.stream_id
                    for out in terminal_outputs]
      for stream_id in set(stream_ids):
        self._add_all_grouping(self.TERMINAL_BOLT_NAME, terminal, stream_id)

    # create topology class
    class_dict = self._construct_topo_class_dict()
    return TopologyType(self.topology_name, (Topology,), class_dict)

  def _add_all_grouping(self, child, parent, stream_id):
    """Adds all grouping between child component and parent component with a given stream id

    :type child: str
    :param child: child's component name
    :type parent: str
    :param parent: parent's component name
    :type stream_id: str
    :param stream_id: stream id
    """
    # child has to be a bolt
    child_component_spec = self.bolts[child]
    # child_inputs is dict mapping from <HeronComponentSpec|GlobalStreamId -> grouping>
    child_inputs = child_component_spec.inputs

    if parent in self.bolts:
      parent_component_spec = self.bolts[parent]
    else:
      parent_component_spec = self.spouts[parent]

    if stream_id == Stream.DEFAULT_STREAM_ID:
      child_inputs[parent_component_spec] = Grouping.ALL
    else:
      child_inputs[parent_component_spec[stream_id]] = Grouping.ALL
